src/route/seg6.rs

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: MIT
+
+use anyhow::Context;
+use byteorder::{ByteOrder, NativeEndian};
+use netlink_packet_utils::{
+    nla::{DefaultNla, Nla, NlaBuffer},
+    Parseable,
+};
+use std::net::Ipv6Addr;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[non_exhaustive]
+pub enum Seg6Mode {
+    // Inline mode for Seg6
+    Inline,
+    // Encapsulation mode for Seg6
+    Encap,
+    // L2ENCAP = 2,
+    // ENCAP_RED = 3,
+    // L2ENCAP_RED = 4
+    Other(u32),
+}
+
+impl From<Seg6Mode> for u32 {
+    fn from(value: Seg6Mode) -> Self {
+        match value {
+            Seg6Mode::Inline => 0,
+            Seg6Mode::Encap => 1,
+            Seg6Mode::Other(i) => i,
+        }
+    }
+}
+
+impl From<u32> for Seg6Mode {
+    fn from(value: u32) -> Self {
+        match value {
+            0 => Seg6Mode::Inline,
+            1 => Seg6Mode::Encap,
+            v => Seg6Mode::Other(v),
+        }
+    }
+}
+
+const SEG6_IPTUNNEL_SRH: u16 = 1;
+
+/// Netlink attributes for `RTA_ENCAP` with `RTA_ENCAP_TYPE` set to
+/// `LWTUNNEL_ENCAP_SEG6`.
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[non_exhaustive]
+pub enum RouteSeg6IpTunnel {
+    // Use an IPv6 segment routing header
+    Seg6(Seg6Header),
+    Other(DefaultNla),
+}
+
+impl Nla for RouteSeg6IpTunnel {
+    fn value_len(&self) -> usize {
+        match self {
+            RouteSeg6IpTunnel::Seg6(v) => v.value_len(),
+            RouteSeg6IpTunnel::Other(v) => v.value_len(),
+        }
+    }
+
+    fn kind(&self) -> u16 {
+        match self {
+            RouteSeg6IpTunnel::Seg6(v) => v.kind(),
+            RouteSeg6IpTunnel::Other(v) => v.kind(),
+        }
+    }
+
+    fn emit_value(&self, buffer: &mut [u8]) {
+        match self {
+            RouteSeg6IpTunnel::Seg6(v) => v.emit_value(buffer),
+            RouteSeg6IpTunnel::Other(v) => v.emit_value(buffer),
+        }
+    }
+}
+
+/// Netlink attributes for `RTA_ENCAP` with `RTA_ENCAP_TYPE` set to
+/// `LWTUNNEL_ENCAP_SEG6`.
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[non_exhaustive]
+pub struct Seg6Header {
+    // Operation mode
+    pub mode: Seg6Mode,
+    // List of segments
+    pub segments: Vec<Ipv6Addr>,
+}
+
+impl Nla for Seg6Header {
+    fn value_len(&self) -> usize {
+        let segments = match self.mode {
+            Seg6Mode::Inline => self.segments.len() + 1,
+            Seg6Mode::Encap => self.segments.len(),
+            Seg6Mode::Other(_) => self.segments.len(),
+        };
+        12 + 16 * segments
+    }
+
+    fn kind(&self) -> u16 {
+        SEG6_IPTUNNEL_SRH
+    }
+
+    fn emit_value(&self, buffer: &mut [u8]) {
+        // Some sources for understanding the format of Seg6 in Netlink
+        // https://github.com/torvalds/linux/blob/master/include/uapi/linux/seg6.h
+        // https://github.com/iproute2/iproute2/blob/main/include/uapi/linux/seg6_iptunnel.h#L27
+        // https://github.com/iproute2/iproute2/blob/e3f9681d4a777fb2595a322b421abf0036ab1aae/ip/iproute_lwtunnel.c#L952
```

Comment:
These files will be outdated especially the ones with line number.

You can remove them or replaced with file name and function name, for example, `iproute2 ip/iproute_lwtunnel.c parse_encap_seg6()`.